internals/overlord/servstate/manager.go

```diff
@@ -75,165 +59,77 @@ func NewManager(s *state.State, runner *state.TaskRunner, pebbleDir string, serv
-func (m *ServiceManager) Plan() (*plan.Plan, error) {
-	releasePlan, err := m.acquirePlan()
-	if err != nil {
-		return nil, err
-	}
-	defer releasePlan()
-	return m.plan, nil
-}
-
-// AppendLayer appends the given layer to the plan's layers and updates the
-// layer.Order field to the new order. If a layer with layer.Label already
-// exists, return an error of type *LabelExists.
-func (m *ServiceManager) AppendLayer(layer *plan.Layer) error {
-	releasePlan, err := m.acquirePlan()
-	if err != nil {
-		return err
-	}
-	defer releasePlan()
-
-	index, _ := findLayer(m.plan.Layers, layer.Label)
-	if index >= 0 {
-		return &LabelExists{Label: layer.Label}
-	}
-
-	return m.appendLayer(layer)
+// PlanChanged informs the service manager that the plan has been updated.
+func (m *ServiceManager) PlanChanged(plan *plan.Plan) {
+	m.planLock.Lock()
+	defer m.planLock.Unlock()
+	m.planUnsafe = plan
 }
 
-func (m *ServiceManager) appendLayer(layer *plan.Layer) error {
-	newOrder := 1
-	if len(m.plan.Layers) > 0 {
-		last := m.plan.Layers[len(m.plan.Layers)-1]
-		newOrder = last.Order + 1
-	}
-
-	newLayers := append(m.plan.Layers, layer)
-	err := m.updatePlanLayers(newLayers)
-	if err != nil {
-		return err
-	}
-	layer.Order = newOrder
-	return nil
+// Plan safely gets a snapshot of the current plan pointer. Plan consumers do
+// not mutate the snapshot, and only react to updates from the Plan Manager.
+func (m *ServiceManager) plan() *plan.Plan {
+	m.planLock.Lock()
+	defer m.planLock.Unlock()
+	// This should never be possible, but lets make the requirements clear to
+	// catch misuse during development. Managers using the plan must receive
+	// a PlanChanged update before the plan is used. The first update will be
+	// received during stateengine StartUp, after the plan manager loads the
+	// plan layers from storage.
+	if m.planUnsafe == nil {
+		panic("service manager with invalid plan state")
+	}
+	return m.planUnsafe
 }
 
-func (m *ServiceManager) updatePlanLayers(layers []*plan.Layer) error {
-	combined, err := plan.CombineLayers(layers...)
-	if err != nil {
-		return err
-	}
-	p := &plan.Plan{
-		Layers:     layers,
-		Services:   combined.Services,
-		Checks:     combined.Checks,
-		LogTargets: combined.LogTargets,
-	}
-	err = p.Validate()
-	if err != nil {
-		return err
+// ServiceArgsLayerCreator defines a map of services and respective command
+// arguments, as provided by "pebble run --args" CLI command. It
+// implements the LayerCreator interface, allowing the type to be used with
+// plan manager's (planstate) AppendLayer or CombineLayer methods, to apply
+// the changes to the plan.
+type ServiceArgsLayerCreator map[string][]string
+
+// Layer presents the requested service argument changes in Layer form, ready
+// to apply to the plan. The planstate lock will be acquired while this method
+// is called, so no additional locking needed.
+func (serviceArgs ServiceArgsLayerCreator) Layer(planIn *plan.Plan) (*plan.Layer, error) {
```

Comment:
This is the example of the ```LayerCreator``` compatible type implemented by the Service Manager. It basically contains the code from ```SetServiceArgs```, but allows the type to be used by the traditional ```AppendLayer``` and ```CombineLayer``` directly, in a way that allows for atomic custom layer creation.

Reply:
Removed.